frontend/gateway/gateway.go

```diff
@@ -738,6 +744,450 @@ func (lbf *llbBridgeForwarder) Inputs(ctx context.Context, in *pb.InputsRequest)
+			Src:      workerRef.ImmutableRef,
+			Selector: m.Selector,
+			Dest:     m.Dest,
+			Readonly: m.Readonly,
+		}
+		if !m.Readonly {
+			ref, err := workerRef.Worker.CacheManager().New(ctx, workerRef.ImmutableRef)
+			if err != nil {
+				return nil, err
+			}
+			go func() {
+				<-release
+				ref.Release(context.TODO())
+			}()
+			execMount.Src = ref
+		}
+		if m.Dest == "/" {
+			ctr.rootFS = execMount.Src
+		} else {
+			ctr.mounts = append(ctr.mounts, execMount)
+		}
+	}
+
+	lbf.ctrsMu.Lock()
+	// ensure we are not clobbering a dup container id request
+	if _, ok := lbf.ctrs[in.ContainerID]; ok {
+		lbf.ctrsMu.Unlock()
+		err = status.Errorf(codes.AlreadyExists, "Container %s already exists", in.ContainerID)
+		return nil, err
+	}
+	lbf.ctrs[in.ContainerID] = ctr
+	lbf.ctrsMu.Unlock()
+	return &pb.NewContainerResponse{}, nil
+}
+
+func (lbf *llbBridgeForwarder) ReleaseContainer(ctx context.Context, in *pb.ReleaseContainerRequest) (*pb.ReleaseContainerResponse, error) {
+	lbf.ctrsMu.Lock()
+	ctr, ok := lbf.ctrs[in.ContainerID]
+	delete(lbf.ctrs, in.ContainerID)
+	lbf.ctrsMu.Unlock()
+	if !ok {
+		return nil, errors.Errorf("Container details for %s not found", in.ContainerID)
+	}
+	close(ctr.release)
+	return &pb.ReleaseContainerResponse{}, nil
+}
+
+type pipeReadWriter struct {
+	r *io.PipeReader
+	w *io.PipeWriter
+}
+
+func (p *pipeReadWriter) Read(b []byte) (int, error) {
+	return p.r.Read(b)
+}
+
+func (p *pipeReadWriter) Write(b []byte) (int, error) {
+	return p.w.Write(b)
+}
+
+func (p *pipeReadWriter) Close() error {
+	rErr := p.r.Close()
+	wErr := p.w.Close()
+	if rErr != nil {
+		return rErr
+	}
+	return wErr
+}
+
+type processIO struct {
+	fds    [3]io.ReadWriteCloser
+	resize chan executor.WinSize
+	done   chan struct{}
+}
+
+func newProcessIO(openFds []uint32) (pio *processIO, err error) {
+	pio = &processIO{
+		resize: make(chan executor.WinSize),
+		done:   make(chan struct{}),
+	}
+
+	defer func() {
+		if err != nil {
+			pio.Close()
+		}
+	}()
+
+	for _, fd := range openFds {
+		// TODO how do we handle non stdio fds?  We will need
+		// to know if they are intended for input or output so we
+		// can proxy bytes in the right direction.  For now only
+		// deal with fd 0, 1, 2
+		if fd <= 2 {
+			r, w := io.Pipe()
+			pio.fds[fd] = &pipeReadWriter{r, w}
+		}
+	}
+
+	return pio, nil
+}
+
+func (pio *processIO) Close() (err error) {
+	close(pio.done)
+	close(pio.resize)
+	// close all the fds but collect the first error to return
+	for _, pipe := range pio.fds {
+		if pipe != nil {
+			closeErr := pipe.Close()
+			if closeErr != nil && err == nil {
+				err = closeErr
+			}
+		}
+	}
+	return err
+}
+
+func (pio *processIO) Resize(ctx context.Context, r *pb.ResizeMessage) {
+	select {
+	case <-ctx.Done():
+	case <-pio.done:
+	default:
+		pio.resize <- executor.WinSize{
+			Cols: r.Cols,
+			Rows: r.Rows,
+		}
+	}
+}
+
+func (pio *processIO) Write(f *pb.FdMessage) error {
+	writer := pio.fds[f.Fd]
```

Comment:
Can multiple clients `ExecProcess` with the same `ProcessID`? If so, we need to protect these map access.

Reply:
Multiple clients will not have access to the same processIO objects, they are stored locally per connection in the `ExecProcess` in the `pios` map (which is guarded within the `ExecProcess` with mutexes)